File: backend/internal/email/email_service.go
package email

import (
	"fmt"
	"net/url"

	"nexia-backend/internal/config"

	resend "github.com/resend/resend-go/v2"
	"go.uber.org/zap"
)

type EmailService struct {
	client      *resend.Client
	fromAddress string
	appBaseURL  string
	enabled     bool
	logger      *zap.Logger
	sendEmail   func(*resend.SendEmailRequest) (*resend.SendEmailResponse, error)
}

func NewEmailService(cfg *config.Config, logger *zap.Logger) *EmailService {
	if cfg.Email.ResendAPIKey == "" {
		logger.Warn("email service disabled: missing resend api key")
		return &EmailService{
			enabled:     false,
			fromAddress: cfg.Email.FromAddress,
			appBaseURL:  cfg.Email.AppBaseURL,
			logger:      logger.Named("email"),
		}
	}
	client := resend.NewClient(cfg.Email.ResendAPIKey)
	return &EmailService{
		client:      client,
		fromAddress: cfg.Email.FromAddress,
		appBaseURL:  cfg.Email.AppBaseURL,
		enabled:     true,
		logger:      logger.Named("email"),
		sendEmail:   client.Emails.Send,
	}
}

func (e *EmailService) SendVerificationEmail(toEmail, token string) error {
	verifyURL := fmt.Sprintf("%s/verify-email/confirm?token=%s", e.appBaseURL, url.QueryEscape(token))

	if !e.enabled {
		e.logger.Info("email send skipped: verification email",
			zap.String("to_email", toEmail),
			zap.String("verify_url", verifyURL),
		)
		return nil
	}

	params := &resend.SendEmailRequest{
		From:    e.fromAddress,
		To:      []string{toEmail},
		Subject: "Verify your Nexia email address",
		Html:    buildVerificationEmailHTML(verifyURL),
	}

	_, err := e.sendEmail(params)
	if err != nil {
		return fmt.Errorf("resend: send verification email: %w", err)
	}
	e.logger.Info("verification email sent", zap.String("to_email", toEmail))
	return nil
}

func (e *EmailService) SendPasswordResetEmail(toEmail, token string) error {
	resetURL := fmt.Sprintf("%s/reset-password", e.appBaseURL)

	if !e.enabled {
		e.logger.Info("email send skipped: password reset email",
			zap.String("to_email", toEmail),
			zap.String("reset_url", resetURL),
		)
		return nil
	}

	params := &resend.SendEmailRequest{
		From:    e.fromAddress,
		To:      []string{toEmail},
		Subject: "Reset your Nexia password",
		Html:    buildPasswordResetEmailHTML(token, resetURL),
	}

	_, err := e.sendEmail(params)
	if err != nil {
		return fmt.Errorf("resend: send password reset email: %w", err)
	}
	e.logger.Info("password reset email sent", zap.String("to_email", toEmail))
	return nil
}

File: apps/api/src/email/email-service.ts
import type { Config } from "../config/config";
import type { Logger } from "../logging/logger";
import { errEmailUnavailable } from "../services/errors";
import { buildVerificationEmailHTML, buildPasswordResetEmailHTML } from "./templates";

const RESEND_ENDPOINT = "https://api.resend.com/emails";

/**
 * Signup blocks on this call, so an unresponsive provider must not hold the
 * request open indefinitely.
 */
const SEND_TIMEOUT_MS = 10_000;

interface SendParams {
  toEmail: string;
  subject: string;
  html: string;
  /** Used in log messages and error text, e.g. "verification email". */
  kind: string;
  /** Extra fields for the "skipped" log line when sending is disabled. */
  skipFields?: Record<string, unknown>;
}

export class EmailService {
  private fromAddress: string;
  private appBaseURL: string;
  private enabled: boolean;
  private apiKey: string;
  private logger: Logger;

  constructor(cfg: Config, logger: Logger) {
    this.fromAddress = cfg.email.from_address;
    this.appBaseURL = cfg.email.app_base_url;
    this.apiKey = cfg.email.resend_api_key;
    this.enabled = Boolean(cfg.email.resend_api_key);
    this.logger = logger.child({ component: "email" });
  }

  async sendVerificationEmail(toEmail: string, token: string): Promise<void> {
    const verifyURL = `${this.appBaseURL}/verify-email/confirm?token=${encodeURIComponent(token)}`;
    await this.send({
      toEmail,
      subject: "Verify your Nexia email address",
      html: buildVerificationEmailHTML(verifyURL),
      kind: "verification email",
      skipFields: { verifyURL },
    });
  }

  async sendPasswordResetEmail(toEmail: string, token: string): Promise<void> {
    const resetURL = `${this.appBaseURL}/reset-password`;
    await this.send({
      toEmail,
      subject: "Reset your Nexia password",
      html: buildPasswordResetEmailHTML(token, resetURL),
      kind: "password reset email",
      skipFields: { resetURL },
    });
  }

  private async send({ toEmail, subject, html, kind, skipFields }: SendParams): Promise<void> {
    if (!this.enabled) {
      // Without a key, log the link so local development can still follow it.
      this.logger.info({ toEmail, ...skipFields }, `email send skipped: ${kind}`);
      return;
    }

    let res: Response;
    try {
      res = await fetch(RESEND_ENDPOINT, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          from: this.fromAddress,
          to: [toEmail],
          subject,
          html,
        }),
        signal: AbortSignal.timeout(SEND_TIMEOUT_MS),
      });
    } catch (err) {
      // Transport-level failure: DNS, TLS, timeout, connection reset.
      throw errEmailUnavailable(
        `resend: send ${kind}: ${err instanceof Error ? err.message : String(err)}`
      );
    }

    // Checked outside the try so a rejected send is not re-wrapped as though it
    // were a transport error.
    if (!res.ok) {
      const body = await res.text();
      throw errEmailUnavailable(`resend: send ${kind}: API error ${res.status} ${body}`);
    }

    this.logger.info({ toEmail }, `${kind} sent`);
  }
}
